File: ERPSystem/ERP.AuthService/Application/Services/AuditLogService.cs
using ERP.AuthService.Application.DTOs.AuditLog;
using ERP.AuthService.Application.DTOs.AuthUser;
using ERP.AuthService.Application.Interfaces.Repositories;
using ERP.AuthService.Application.Interfaces.Services;
using ERP.AuthService.Domain.Logger;

namespace ERP.AuthService.Application.Services
{
    public class AuditLogService : IAuditLogService
    {
        private readonly IAuditLogRepository _repository;
        private readonly ITenantContext _tenantContext;

        public AuditLogService(IAuditLogRepository repository, ITenantContext tenantContext)
        {
            _repository = repository;
            _tenantContext = tenantContext;
        }

        /// <summary>Get all audit logs, paginated, sorted by most recent.</summary>
        public async Task<PagedResultDto<AuditLogResponseDto>> GetAllAsync(int pageNumber, int pageSize)
        {
            List<AuditLog> items = await _repository.GetAllAsync(pageNumber, pageSize);
            long total = await _repository.CountAsync();
            return new PagedResultDto<AuditLogResponseDto>(
                items.Select(MapToDto).ToList(),
                (int)total,
                pageNumber,
                pageSize);
        }

        /// <summary>Get all logs where the user is either the performer or the target.</summary>
        public async Task<PagedResultDto<AuditLogResponseDto>> GetByUserAsync(Guid userId, int pageNumber, int pageSize)
        {
            List<AuditLog> items = await _repository.GetByUserAsync(userId, pageNumber, pageSize);
            return new PagedResultDto<AuditLogResponseDto>(
                items.Select(MapToDto).ToList(),
                items.Count,
                pageNumber,
                pageSize);
        }

        /// <summary>Get total count of all audit log entries.</summary>
        public async Task<long> CountAsync()
            => await _repository.CountAsync();

        /// <summary>Clear all audit logs (development only).</summary>
        public async Task ClearAsync()
            => await _repository.ClearAsync();

        // ── Mapping ───────────────────────────────────────────────────────────

        private static AuditLogResponseDto MapToDto(AuditLog log) =>
            new(
                log.Id,
                log.Action,
                log.PerformedBy,
                log.TargetUserId,
                log.Success,
                log.FailureReason,
                log.IpAddress,
                log.UserAgent,
                log.Metadata,
                log.Timestamp
            );
    }
}

File: ERP.Frontend/src/app/services/audit-log.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../environment';

// ─── DTOs ─────────────────────────────────────────────────────────────────────

export type AuditAction =
  // Auth
  | 'Login'
  | 'Logout'
  | 'TokenRefreshed'
  | 'TokenRevoked'
  | 'TokenValidated'           // ✅ new
  | 'TokenValidationFailed'    // ✅ new

  // Registration
  | 'UserRegistered'

  // Password
  | 'PasswordChanged'
  | 'PasswordChangedByAdmin'

  // Profile
  | 'ProfileUpdated'

  // Account status
  | 'UserActivated'
  | 'UserDeactivated'
  | 'UserDeleted'
  | 'UserRestored'

  // Role
  | 'RoleCreated'
  | 'RoleUpdated'
  | 'RoleDeleted'

  // Controle
  | 'ControleCreated'
  | 'ControleUpdated'
  | 'ControleDeleted'

  | 'Unauthorized'
  | 'UserNotFound'
  | 'UnhandledError';

export interface AuditLogResponseDto {
  id: string;
  action: AuditAction;
  performedBy: string | null;
  targetUserId: string | null;
  success: boolean;
  failureReason: string | null;
  ipAddress: string | null;
  userAgent: string | null;
  metadata: Record<string, string> | null;
  timestamp: string;
}

export interface PagedResultDto<T> {
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
}

// ─── Service ──────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class AuditLogService {
  private readonly baseUrl = `/audit`;

  constructor(private http: HttpClient) {}

  /** GET /auth/audit — Get all logs paginated */
  getAll(pageNumber = 1, pageSize = 20): Observable<PagedResultDto<AuditLogResponseDto>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber)
      .set('pageSize', pageSize);
    return this.http.get<PagedResultDto<AuditLogResponseDto>>(this.baseUrl, { params });
  }

  /** GET /auth/audit/user/{userId} — Get logs by user (performer or target) */
  getByUser(userId: string, pageNumber = 1, pageSize = 20): Observable<PagedResultDto<AuditLogResponseDto>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber)
      .set('pageSize', pageSize);
    return this.http.get<PagedResultDto<AuditLogResponseDto>>(
      `${this.baseUrl}/user/${userId}`, { params });
  }

  /** GET /auth/audit/count — Get total audit log count */
  count(): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/count`);
  }

  /** DELETE /auth/audit — Clear all audit logs (dev only) */
  clear(): Observable<void> {
    return this.http.delete<void>(this.baseUrl);
  }
}
